fe/fe-core/src/main/java/org/apache/doris/catalog/ColumnType.java

```diff
@@ -208,12 +210,48 @@ public static void checkSupportSchemaChangeForComplexType(Type checkType, Type o
         if (checkType.isStructType() && other.isStructType()) {
             StructType thisStructType = (StructType) checkType;
             StructType otherStructType = (StructType) other;
-            if (thisStructType.getFields().size() != otherStructType.getFields().size()) {
-                throw new DdlException("Cannot change struct type with different field size");
+            // now we only support add new field for struct type
+            if (thisStructType.getFields().size() > otherStructType.getFields().size()) {
+                throw new DdlException("Cannot reduce struct fields from " + checkType.toSql() + " to "
+                        + other.toSql());
             }
-            for (int i = 0; i < thisStructType.getFields().size(); i++) {
+            int i = 0;
+            HashSet<String> thisStructNames = new HashSet<>();
+            for (; i < thisStructType.getFields().size(); i++) {
+                // do not support struct same position field name/type change
+                StructField thisField = thisStructType.getFields().get(i);
+                StructField otherField = otherStructType.getFields().get(i);
+                if (!thisField.equals(otherField)) {
+                    // name and type are changed will throw exception
+                    if (!thisField.getName().equals(otherField.getName())) {
+                        throw new DdlException("Cannot change name in struct" + thisField.getName() + " to "
+                                + otherField.getName());
+                    }
+                    if (!thisField.getType().equals(otherField.getType())) {
+                        if (thisField.getType().isStringType() && otherField.getType().isStringType()) {
+                            // check string type
+                            // check length growth later in checkSupportSchemaChangeForCharType
+                            checkSupportSchemaChangeForCharType(thisStructType.getFields().get(i).getType(),
+                                    otherStructType.getFields().get(i).getType());
+                        } else if (thisField.getType().isComplexType() && otherField.getType().isComplexType()) {
+                            // check complex type recursively
+                            checkSupportSchemaChangeForComplexType(thisStructType.getFields().get(i).getType(),
+                                    otherStructType.getFields().get(i).getType(), true);
+                        } else {
+                            throw new DdlException("Cannot change " + checkType.toSql() + " to " + other.toSql());
+                        }
+                    }
+                }
                 checkSupportSchemaChangeForComplexType(thisStructType.getFields().get(i).getType(),
```

Comment:
no need  to call `checkSupportSchemaChangeForComplexType`  when `thisField.equals(otherField)`